sdk-extensions/tracing-incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/trace/LeakDetectingSpanProcessor.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.extension.incubator.trace;
+
+import static java.lang.Thread.currentThread;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.internal.shaded.WeakConcurrentMap;
+import io.opentelemetry.sdk.trace.ReadWriteSpan;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import java.lang.ref.Reference;
+import java.util.Arrays;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A {@link SpanProcessor} which will detect spans that are never ended. It will detect spans that
+ * are garbage collected without ever having `end()` called on them.
+ *
+ * <p>Note: using this SpanProcessor will definitely impact the performance of your application. It
+ * is not recommended for production use, as it uses additional memory for each span to track where
+ * a leaked span was created.
+ */
+public final class LeakDetectingSpanProcessor implements SpanProcessor {
+  private static final Logger logger = Logger.getLogger(LeakDetectingSpanProcessor.class.getName());
+
+  private final PendingSpans pendingSpans;
+
+  /**
+   * Create a new {@link LeakDetectingSpanProcessor} that will report any un-ended spans that get
+   * garbage collected.
+   */
+  public static LeakDetectingSpanProcessor create() {
+    return new LeakDetectingSpanProcessor();
+  }
+
+  private LeakDetectingSpanProcessor() {
```

Comment:
Ah - don't think we need this constructor since we have the factory :)

Reply:
fair enough!

Reply:
inlined